File: tools/seqlen_stats.py
#!/usr/bin/env python3
"""Measure the token-length distribution of an MLX chat-format JSONL dataset.

Usage: seqlen_stats.py dataset.jsonl model_dir [max_len]

Reports sample count, min/mean/p50/p90/p99/max token lengths (chat-template
applied, exactly like mlx-lm does), and how many samples exceed max_len
(default 4096) — i.e. how much of the data gets truncated during training.
"""

import json
import sys
from transformers import AutoTokenizer


def main() -> None:
    path, model_dir = sys.argv[1], sys.argv[2]
    max_len = int(sys.argv[3]) if len(sys.argv) > 3 else 4096

    tok = AutoTokenizer.from_pretrained(model_dir)
    lens = []
    bad = 0
    for line in open(path, encoding="utf-8"):
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
            msgs = obj.get("messages")
            if msgs is None:
                # ShareGPT format: {"conversations": [{from, value}, ...]}
                msgs = [
                    {"role": "user" if c["from"] == "human" else "assistant",
                     "content": c["value"]}
                    for c in obj["conversations"]
                ]
            text = tok.apply_chat_template(msgs, tokenize=False)
            lens.append(len(tok.encode(text)))
        except Exception as e:
            bad += 1
            print(f"[warn] skipping entry: {e}")

    lens.sort()
    n = len(lens)
    if n == 0:
        print("no samples")
        return

    def pct(p):
        return lens[min(n - 1, int(n * p))]

    over = sum(1 for l in lens if l > max_len)
    print(f"samples: {n} (skipped {bad})")
    print(f"tokens:  min {lens[0]}  mean {sum(lens) // n}  "
          f"p50 {pct(0.5)}  p90 {pct(0.9)}  p99 {pct(0.99)}  max {lens[-1]}")
    print(f"> {max_len} tokens (truncated during training): {over} "
          f"({over / n * 100:.1f}%)")
    if over:
        print("longest few: " + ", ".join(str(l) for l in lens[-5:]))


if __name__ == "__main__":
    main()

File: tools/seqlen_stats.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Measure the token-length distribution of a dataset (diagnostic helper).
#
# Reports sample count, min/mean/p50/p90/p99/max lengths, and how many
# samples exceed max_len — i.e. how much of the data gets truncated during
# training.
#
# Two counting modes:
#   - with a model directory: the real tokenizer (tools/tokenizer.rb — byte-
#     level BPE + chat template, verified identical to the original
#     Python/transformers tokenizer on this pipeline's data)
#   - without: the generator's conservative estimator (SPLIT_CHARS_PER_TOKEN
#     in build_dataset.rb, ~1.4 chars/byte + chat-template overhead)
#
# Accepts both ShareGPT {"conversations": [...]} and MLX {"messages": [...]}.
#
# Usage:
#   ruby tools/seqlen_stats.rb dataset.jsonl [max_len]
#   ruby tools/seqlen_stats.rb dataset.jsonl <model_dir> [max_len]
#   (default max_len = MAX_CONTEXT_TOKENS)

require "json"
require_relative "../build/build_dataset"
require_relative "tokenizer"

def messages_from(obj)
  if obj["messages"]
    obj["messages"]
  else
    obj["conversations"].map do |c|
      { "role" => c["from"] == "human" ? "user" : "assistant", "content" => c["value"] }
    end
  end
end

def entry_length(msgs, tokenizer)
  return tokenizer.count_messages(msgs) if tokenizer

  estimate_tokens(msgs.map { |m| m["content"].to_s }.join("\n")) + CHAT_TEMPLATE_OVERHEAD_TOKENS
end

# Collects the length of every parseable line of `lines` (any enumerable of
# JSONL lines). Returns [lens, skipped_count].
def collect_lengths(lines, tokenizer: nil)
  lens = []
  bad = 0
  lines.each_with_index do |line, i|
    next if line.strip.empty?

    begin
      msgs = messages_from(JSON.parse(line))
    rescue JSON::ParserError => e
      bad += 1
      warn "  [warn] line #{i + 1}: #{e.message}"
      next
    end

    lens << entry_length(msgs, tokenizer)
  end
  [lens, bad]
end

# Computes the distribution stats from a SORTED lens array (must not be
# empty). Returns a hash with samples/min/mean/p50/p90/p99/max/over/
# over_pct/label.
def summarize(lens, max_len, real:)
  n = lens.size
  pct = lambda do |p|
    lens[[(n * p).to_i, n - 1].min]
  end
  over = lens.count { |l| l > max_len }
  {
    samples: n,
    min: lens.first, mean: lens.sum / n,
    p50: pct.call(0.5), p90: pct.call(0.9), p99: pct.call(0.99), max: lens.last,
    over: over, over_pct: over * 100.0 / n,
    label: real ? "tokens" : "est tokens",
  }
end

if __FILE__ == $PROGRAM_NAME
  PATH = ARGV[0]
  abort "usage: #{$PROGRAM_NAME} dataset.jsonl [model_dir] [max_len]" if PATH.nil? || !File.file?(PATH)

  REST = ARGV[1..]
  MODEL_DIR = REST.find { |a| File.directory?(a) }
  MAX_LEN = REST.find { |a| a =~ /\A\d+\z/ }&.to_i || MAX_CONTEXT_TOKENS

  TOKENIZER = MODEL_DIR ? Tokenizer.new(MODEL_DIR) : nil

  lens, bad = collect_lengths(File.foreach(PATH, encoding: "UTF-8"), tokenizer: TOKENIZER)
  lens.sort!
  if lens.empty?
    puts "no samples"
    exit 0
  end

  s = summarize(lens, MAX_LEN, real: !TOKENIZER.nil?)
  puts "samples: #{s[:samples]} (skipped #{bad})"
  puts format("#{s[:label]}: min %d  mean %d  p50 %d  p90 %d  p99 %d  max %d",
              s[:min], s[:mean], s[:p50], s[:p90], s[:p99], s[:max])
  puts format("> %d #{s[:label]} (truncated during training): %d (%.1f%%)",
              MAX_LEN, s[:over], s[:over_pct])
  puts "longest few: #{lens.last(5).join(', ')}" if s[:over].positive?
end
